Introducing NumericInput Component

shaavan opened the pull request:
Resolves #443

- This PR introduces the `NumericInput` component.
- Allow increment or decrement of the number based on the step size
- Contains all the relevant properties of the `<input type="number">`
- Can be used as both _controlled_ and _uncontrolled_ components based on the provided props.

shaavan commented:
Right now, this PR introduces the skeleton of the component without the relevant styling, which will be addressed in a follow-up commit.

Also, this PR has introduced the `NumericInput` bar at the top of lesson content for easier testing by reviewers. This, too, shall be cleaned up in a follow-up commit.

@afterburn 

Please have a look through the code and functionality, and confirm if it satisfies our needs for the `NumericInput` component.

afterburn commented on api/local/getProgress.ts:
```diff
@@ -2,12 +2,14 @@ import { defaultProgressContext } from 'providers/ProgressProvider'
       throw new Error('Could not read progress from LocalStorage')
     }
 
+    const res = progress.replace(/['"]/g, '')
```
Wasn't this removed in a previous PR?

shaavan replied:
Yes, this is fixed in #452.

This commit was a temporary fix to avoid crashing PR due to an unrelated bug.
I will remove this with the next update of PR.

shaavan replied:
Addressed in 1c7cd6cb913ce1305fa02df9af0a4e5b9c955001

Thanks!

afterburn commented on components/Util/NumericInput.tsx:
```diff
@@ -0,0 +1,40 @@
+  step?: number
+  defaultValue?: number
+  value?: number
+  onChange?: (evt: any) => void
```
Let's make it so this component returns a `number`. To me it makes sense to handle the event itself inside this component and only pass the resulting number up to the parent.  If I choose to use a `NumericInput` I expect the value that comes from it to be a number.

shaavan replied:
Thanks for the review @afterburn.

I have some doubts to make about your comment.

> To me it makes sense to handle the event itself inside this component

Wouldn't handling the event inside the component make it an _uncontrolled component_? By definition, _controlled component_ states are governed by their parents.
I have allowed the possibility of both based on the description in #443 (For engineers, point 2).

afterburn replied:
What I suggest is that we handle the event inside the component, this does not make it an uncontrolled component because as a developer you can still choose to manage the state for the input in the parent or have the `NumericInput` track state which is what the difference is between a controlled and uncontrolled component, where we handle the event has nothing to do with that.

Something like this:
```tsx
function NumericInput() {
  const [value, setValue] = useState(value)

  const handleChange = e => {
    const newValue = Number(e.target.value)
    
    // Based on whether its a controlled or uncontrolled component:
    if (controlled) {
       onChange(newValue)
    } else {
      setValue(newValue)
    }
  }

  return <input onChange={handleChange} />
}
```

This makes the component easier to use, which is the point:

```
function Parent() {
  const handleChange = (val: number) => {
    console.log(val) // Outputs number
  }
  
  return <NumericInput onChange={handleChange} />
}
```

afterburn replied:
Full example of what I mean [here](https://codesandbox.io/s/numeric-input-xgpd6t)

shaavan replied:
Thanks, @afterburn, for the example. This helps clears up my doubts!

Let me take references from it to create the final component!

shaavan replied:
Addressed in 1c7cd6cb913ce1305fa02df9af0a4e5b9c955001
I have taken reference from your example to create the final component.
Thanks for the detailed example!

afterburn commented on app/[lang]/chapters/[slug]/[lesson]/page.tsx:
```diff
@@ -112,10 +115,19 @@ export default function Page({ params }) {
+        <NumericInput
+          label={'string'}
+          value={val}
+          onChange={(e) => setVal(e.target.value)}
```
I think `e.target.value` will actually be a string in some browsers, we need to make sure it returns a number (see prev comments)

shaavan replied:
Addressed in 1c7cd6cb913ce1305fa02df9af0a4e5b9c955001
The latest implementation passes the setValue function in the NumericInput.

shaavan commented on components/Util/NumericInput.tsx:
```diff
@@ -0,0 +1,40 @@
+      defaultValue={defaultValue ?? undefined}
+      //For Controlled Component
+      value={value ?? undefined}
+      onChange={onChange ?? undefined}
```
Out of all events, `onChange` seems the most logical one to handle first.
The other events can be handled later, as they will become relevant for the `NumericInput` component.

afterburn replied:
I feel we also want to handle `onInput` at least

shaavan replied:
According to [w3schools](https://www.w3schools.com/jsref/event_oninput.asp#:~:text=The%20oninput%20event%20is%20similar,when%20the%20element%20loses%20focus.)

> The oninput event is similar to the [onchange](https://www.w3schools.com/jsref/event_onchange.asp) event.
The difference is that the oninput event occurs immediately after the content has been changed, while onchange occurs when the element loses focus.

Since they are so similar in behavior, tackling both might unnecessarily increase complexity.

vinayaksh42 commented:
I am unable to see the `Uncontrolled` NumericInput component in Vercel deployment link, although it is visible when I am running this PR locally. 
<img width="392" alt="Screenshot 2023-05-29 at 2 18 17 PM" src="https://github.com/saving-satoshi/saving-satoshi/assets/54861487/3b8bcc00-b293-46bf-bd92-0e839f693718">

vinayaksh42 commented on components/Util/NumericInput.tsx:
```diff
@@ -0,0 +1,63 @@
+  }, [isControlled, value])
+
+  return (
+    <input
```
I think we should add onkeydown for the enter key, the same way we have in ui/lesson/Terminal.tsx.

shaavan replied:
Seems reasonable to me.

Since we don't have any place to accept the numericInput, let me create a dummy handler that logs the value in the terminal when "Enter" is pressed.

shaavan replied:
Also @vinayaksh42 

> I am unable to see the Uncontrolled NumericInput component in Vercel deployment link

I opened the Vercel preview [link](https://saving-satoshi-h2m4c54u1-savingsatoshi.vercel.app/) (Mac, Chrome) in both new and incognito windows and can see both components.

<img width="1507" alt="image" src="https://github.com/saving-satoshi/saving-satoshi/assets/85434418/dedf10b9-01bd-461b-b905-c8d491be80a0">

---

Can it be browser specific problem? Which browser are you working with?

shaavan commented:
Updated from 1c7cd6cb913ce1305fa02df9af0a4e5b9c955001 to 704c28130b7e87852932dc490720f50780f9e2a2 ([pr453.02](https://github.com/shaavan/saving-satoshi/commits/pr453.02) -> [pr453.03](https://github.com/shaavan/saving-satoshi/commits/pr453.03), [diff](https://github.com/shaavan/saving-satoshi/compare/pr453.02..pr453.03))
Addressed @vinayaksh42 comment

- Introduced onKeyDown event in the `NumericInput` component.

afterburn commented on components/Util/NumericInput.tsx:
```diff
@@ -0,0 +1,76 @@
+
+  const isControlled = value !== undefined && onChange !== undefined
+
+  const [internalValue, setInternelValue] = useState(
```
internel -> internal

shaavan replied:
Addressed in cc34c09e6c252ca8d298e44167934c24bd670b70
Thanks!

afterburn commented on components/Util/NumericInput.tsx:
```diff
@@ -0,0 +1,76 @@
+
+  function handleKeyDown(evt) {
+    if (evt.key == 'Enter') {
+      if (isControlled && onKeyDown) {
```
I should also be able to pass an `onKeyDown` to an uncontrolled component in case I want to hook into that event, same goes for `onChange` the only difference between the controlled and uncontrolled variants is how we handle the internal state.

shaavan replied:
Addressed in cc34c09e6c252ca8d298e44167934c24bd670b70
Thanks!

shaavan commented:
Updated from cc34c09e6c252ca8d298e44167934c24bd670b70 to 4022862094a473147c9092f6fb5753def89be919  ([pr453.04](https://github.com/shaavan/saving-satoshi/commits/pr453.04) -> [pr453.05](https://github.com/shaavan/saving-satoshi/commits/pr453.05), [diff](https://github.com/shaavan/saving-satoshi/compare/pr453.04..pr453.05))
Addressed @afterburn comment

Update

- Removed the test code for NumericInput from the page.tsx
